pkg/controller/deployment/deployment_controller.go

```diff
@@ -355,6 +371,44 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 		return err
 	}
 
+	if dc.garbageCollectorEnabled {
+		rsList, err := dc.rsLister.ReplicaSets(deployment.Namespace).List(labels.Everything())
+		if err != nil {
+			glog.Errorf("Error getting ReplicaSets for deployment %q: %v", deployment, err)
+			return err
+		}
+
+		deploymentSelector, err := unversioned.LabelSelectorAsSelector(deployment.Spec.Selector)
+		if err != nil {
+			return fmt.Errorf("deployment %s/%s has invalid label selector: %v", deployment.Namespace, deployment.Name, err)
+		}
+		cm := controller.NewReplicaSetControllerRefManager(dc.rsControl, deployment.ObjectMeta, deploymentSelector, getDeploymentControllerKind())
+		matchesAndControlled, matchesNeedsController, controlledDoesNotMatch := cm.Classify(rsList)
+		for _, replicaSet := range matchesNeedsController {
+			err := cm.AdoptReplicaSet(replicaSet)
+			// continue to next RS if adoption fails.
+			if err != nil {
+				// If the RS no longer exists, don't even log the error.
+				if !errors.IsNotFound(err) {
+					utilruntime.HandleError(err)
+				}
+			} else {
+				matchesAndControlled = append(matchesAndControlled, replicaSet)
+			}
+		}
+		// remove the controllerRef for the RS that no longer have matching labels
+		var errlist []error
+		for _, replicaSet := range controlledDoesNotMatch {
+			err := cm.ReleaseReplicaSet(replicaSet)
+			if err != nil {
+				errlist = append(errlist, cm.ReleaseReplicaSet(replicaSet))
+			}
+		}
+		if len(errlist) != 0 {
+			return utilerrors.NewAggregate(errlist)
+		}
+
+	}
```

Comment:
I don't know deployement controller well enough. If the number of controlled replicaset is greater than the max rollback history, will another routine try to delete the surplus? @janetkuo @kargakis